pkg/ddevapp/config_test.go

```diff
@@ -1317,3 +1319,220 @@ func TestDatabaseConfigUpgrade(t *testing.T) {
 		assert.Empty(app.MariaDBVersion)
 	}
 }
+
+// TestConfigMergeStringList verifies that string merges update w/o clobber
+func TestConfigMergeStringList(t *testing.T) {
+	assert := asrt.New(t)
+	pwd, _ := os.Getwd()
+
+	projDir, err := filepath.Abs(testcommon.CreateTmpDir(t.Name()))
+	require.NoError(t, err)
+
+	err = fileutil.CopyDir("./testdata/TestConfigMergeItems/.ddev", filepath.Join(projDir, ".ddev"))
+	require.NoError(t, err)
+
+	app, err := NewApp(projDir, true)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		err = os.Chdir(pwd)
+		assert.NoError(err)
+		err = app.Stop(true, false)
+		assert.NoError(err)
+		err = os.RemoveAll(projDir)
+		assert.NoError(err)
+	})
+	err = os.Chdir(app.AppRoot)
+	assert.NoError(err)
+
+	// test matches allowing for delete syntax (a prefixed !)
+	assertSimpleMatch := func(expected bool, match string, setting []string) {
+		test := assert.True
+		if !expected {
+			test = assert.False
+		}
+
+		for _, val := range setting {
+			if val == match {
+				test(true, match)
+				return
+			}
+		}
+		test(false, match)
+	}
+	// no clobber of old setting
+	assertSimpleMatch(true, "somename", app.AdditionalHostnames)
+	// successful merge
+	assertSimpleMatch(true, "somename-new", app.AdditionalHostnames)
+	// do not want the delete instruction added to the list
+	assertSimpleMatch(false, "!someothername", app.AdditionalHostnames)
+
+}
+
+// TestConfigMergeEnvItems verifies that config overrides for web_environment
+// override and do not clobber settings from config.yaml.
+func TestConfigMergeEnvItems(t *testing.T) {
+	assert := asrt.New(t)
+	pwd, _ := os.Getwd()
+
+	projDir, err := filepath.Abs(testcommon.CreateTmpDir(t.Name()))
+	require.NoError(t, err)
+
+	err = fileutil.CopyDir("./testdata/TestConfigMergeItems/.ddev", filepath.Join(projDir, ".ddev"))
+	require.NoError(t, err)
+
+	app, err := NewApp(projDir, true)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		err = os.Chdir(pwd)
+		assert.NoError(err)
+		err = app.Stop(true, false)
+		assert.NoError(err)
+		err = os.RemoveAll(projDir)
+		assert.NoError(err)
+	})
+	err = os.Chdir(app.AppRoot)
+	assert.NoError(err)
+
+	// make test results a bit clearer with our
+	// own matcher.
+	assertHasKey := func(expected bool, match string) {
```

Comment:
I think it might be more correct to just provide a testdata config.yaml file with everything included in it and compare it to the merged result. Wouldn't that be more explicit that just listing keys?

Reply:
Typically test data will have a few items in them; the assert wrapper is also munging the data structure, since there's no assert in your library (AFAICT) that does a "string in slice" test.  Keeps the test DRY.